pkg/serving/v1/apply.go

```diff
@@ -0,0 +1,268 @@
+	// Get the current annotations from the object.
+	annots := service.Annotations
+	if annots == nil {
+		annots = map[string]string{}
+	}
+
+	original := annots[v1.LastAppliedConfigAnnotation]
+	delete(annots, v1.LastAppliedConfigAnnotation)
+	service.Annotations = annots
+
+	uModifiedService, err := encodeService(service)
+	if err != nil {
+		return nil, err
+	}
+
+	if annotate {
+		annots[v1.LastAppliedConfigAnnotation] = strings.TrimRight(string(uModifiedService), "\n")
+
+		service.Annotations = annots
+		uModifiedService, err = encodeService(service)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Restore the object to its original condition.
+	annots[v1.LastAppliedConfigAnnotation] = original
+	service.Annotations = annots
+	return uModifiedService, nil
+}
+
+func updateLastAppliedAnnotation(service *servingv1.Service) error {
+	annots := service.Annotations
+	if annots == nil {
+		annots = map[string]string{}
+	}
+	lastApplied, err := encodeService(service)
+	if err != nil {
+		return err
+	}
+
+	// Cleanup any trailing newlines
+	annots[v1.LastAppliedConfigAnnotation] = strings.TrimRight(string(lastApplied), "\n")
+
+	service.Annotations = annots
+	return nil
+}
+
+func encodeService(service *servingv1.Service) ([]byte, error) {
+	scheme := runtime.NewScheme()
+	err := servingv1.AddToScheme(scheme)
+	if err != nil {
+		return nil, err
+	}
+	factory := serializer.NewCodecFactory(scheme)
+	encoder := factory.EncoderForVersion(unstructured.UnstructuredJSONScheme, servingv1.SchemeGroupVersion)
+	err = util.UpdateGroupVersionKindWithScheme(service, servingv1.SchemeGroupVersion, scheme)
+	if err != nil {
+		return nil, err
+	}
+
+	serviceUnstructured, err := util.ToUnstructured(service)
+	if err != nil {
+		return nil, err
+	}
+
+	// Remove/adapt service so that it can be used in the apply-annotation
+	cleanupServiceUnstructured(serviceUnstructured)
+
+	return runtime.Encode(encoder, serviceUnstructured)
+}
+
+func cleanupServiceUnstructured(uService *unstructured.Unstructured) {
+	clearCreationTimestamps(uService.Object)
+	removeStatus(uService.Object)
+	removeContainerNameAndResourcesIfNotSet(uService.Object)
+
+}
+
+func removeContainerNameAndResourcesIfNotSet(uService map[string]interface{}) {
+	uContainer := extractUserContainer(uService)
+	if uContainer == nil {
+		return
+	}
+	name, ok := uContainer["name"]
+	if ok && name != "" {
+		delete(uContainer, "name")
+	}
+
+	resources := uContainer["resources"]
+	if resources == nil {
+		return
+	}
+	resourcesMap := resources.(map[string]interface{})
+	if len(resourcesMap) == 0 {
+		delete(uContainer, "resources")
+	}
+}
+
+func extractUserContainer(uService map[string]interface{}) map[string]interface{} {
+	tSpec := extractTemplateSpec(uService)
+	if tSpec == nil {
+		return nil
+	}
+	containers := tSpec["containers"]
+	if len(containers.([]interface{})) == 0 {
+		return nil
+	}
+	return containers.([]interface{})[0].(map[string]interface{})
```

Comment:
I am not sure about the position of the user-container in the array, is it always at 0 index (thinking about multi-container)?

Reply:
For the moment we are assuming only single, container, that's true. Maybe we should iterate over the list and just pick the container with name `user-container` ?

Reply:
Should this only remove the name of the user-container? or for all the containers in array? If its later, we can simply range over the array and remove the names. I need to catch up on the multi container support to understand this better.